pkg/trimaran/peaks/peaks.go

```diff
@@ -0,0 +1,231 @@
+
+	"sigs.k8s.io/scheduler-plugins/apis/config"
+	cfgv1 "sigs.k8s.io/scheduler-plugins/apis/config/v1"
+	"sigs.k8s.io/scheduler-plugins/pkg/trimaran"
+)
+
+const (
+	Name = "Peaks"
+)
+
+type Peaks struct {
+	handle    framework.Handle
+	collector *trimaran.Collector
+	args      *config.PeaksArgs
+}
+
+type PowerModel struct {
+	K0 float64 `json:"k0"`
+	K1 float64 `json:"k1"`
+	K2 float64 `json:"k2"`
+	// Power = K0 + K1 * e ^(K2 * x) : where x is utilisation
+	// Idle power of node will be K0 + K1
+}
+
+var _ framework.ScorePlugin = &Peaks{}
+var maxPower = 0.0
+var clusterPowerModel map[string]PowerModel
+
+func (pl *Peaks) Name() string {
+	return Name
+}
+
+func initNodePowerModels() error {
+	data, err := os.ReadFile(os.Getenv("NODE_POWER_MODEL"))
+	if err != nil {
+		klog.ErrorS(err, "Unable to read power model from the input configuration")
+		return err
+	}
+	if err = json.Unmarshal(data, &clusterPowerModel); err != nil {
+		klog.ErrorS(err, "Unable to unmarshal power model from the input configuration")
+		return err
+	}
+	return nil
+}
+
+func New(ctx context.Context, obj runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+	logger := klog.FromContext(ctx)
+	klog.V(4).InfoS("Peaks plugin Input config %+v", obj)
+
+	args, ok := obj.(*config.PeaksArgs)
+	if !ok {
+		return nil, fmt.Errorf("want args to be of type PeaksArgs, got %T", obj)
+	}
+	collector, err := trimaran.NewCollector(logger, &config.TrimaranSpec{WatcherAddress: args.WatcherAddress})
+	if err != nil {
+		return nil, err
+	}
+	err = initNodePowerModels()
+	if err != nil {
+		return nil, err
+	}
+	pl := &Peaks{
+		handle:    handle,
+		collector: collector,
+		args:      args,
+	}
+	return pl, nil
+}
+
+func (pl *Peaks) Score(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	logger := klog.FromContext(ctx)
+	score := framework.MinNodeScore
+
+	nodeInfo, err := pl.handle.SnapshotSharedLister().NodeInfos().Get(nodeName)
+	if err != nil {
+		return score, framework.NewStatus(framework.Error, fmt.Sprintf("getting node %q from Snapshot: %v", nodeName, err))
+	}
+
+	metrics, _ := pl.collector.GetNodeMetrics(logger, nodeName)
+	if metrics == nil {
+		klog.ErrorS(nil, "Failed to get metrics for node; using minimum score", "nodeName", nodeName)
+		return score, nil
+	}
+
+	var curPodCPUUsage int64
+	for _, container := range pod.Spec.Containers {
+		curPodCPUUsage += PredictUtilisation(&container)
+	}
+	if pod.Spec.Overhead != nil {
+		curPodCPUUsage += pod.Spec.Overhead.Cpu().MilliValue()
+	}
+
+	var nodeCPUUtilPercent float64
+	var cpuMetricFound bool
+	for _, metric := range metrics {
+		if metric.Type == watcher.CPU {
+			if metric.Operator == watcher.Average || metric.Operator == watcher.Latest {
+				nodeCPUUtilPercent = metric.Value
+				cpuMetricFound = true
+			}
+		}
+	}
+	if !cpuMetricFound {
+		klog.ErrorS(nil, "Cpu metric not found in node metrics for nodeName", nodeName)
```

Comment:
Ditto. Eliminate klog by using contextual logging.

Reply:
Changed to contextual logging.